Flexible tutorials

Arcadius19 opened the pull request:
Tutorials are now dynamically fetched from the server using tutorial API.
Users have been added privileges to distinguish between admins and normal users.
Only admins can create a new tutorial.
Added db-init.sql file which describes a current database schema.
Added ESLint config to make code styling consistent.

Neglect api/compile.js and app.js as these are commits from the previous branch.

Arcadius19 committed:
Changed the allowOrigin array
Fixed the bug which did not allow multiple users at the same time

Arcadius19 committed:
Tutorials' description retrieved from the server automatically.
Implemented POST request handler for the creation of a new tutorial

Arcadius19 committed:
Reverted the list of previous allowed origins with a new server's IP added

Arcadius19 committed:
Merge branch 'AllowConcurrentUsers'

# Conflicts:
#	app.js

Arcadius19 committed:
Created environment variables for allowed origins, Links configuration file and time of the Links code compilation

Arcadius19 committed:
Merge branch 'AllowConcurrentUsers'

Arcadius19 committed:
Changed the storage of tutorials from files to database entries in LinksTutorial table

Arcadius19 committed:
Headers are fetched from the server dynamically

Arcadius19 committed:
Initial source code is retrieved dynamically if requested for the first time (previous version threw an error upon a first retrieval)

Arcadius19 committed:
Users are assigned the rights (usual user or admin).
Upon the login, only user's username and last tutorial id are sent back (previously also e-mail address and password hash)

Arcadius19 committed:
Created ESLint config file and adjusted tutorial.js.
Fixed attribute names in user details sent back upon the login

Arcadius19 committed:
Added authentication for creation of a new tutorial

Arcadius19 committed:
Adjusted the code to ESLint config

Arcadius19 committed:
Added a function to automatically populate initial DB

Arcadius19 committed:
Merge remote-tracking branch 'origin/master'

Arcadius19 committed:
Merge branch 'master' into FlexibleTutorials

NickWu007 commented on api/file.js:
```diff
@@ -14,16 +14,20 @@ function readFile (req, res, next) {
         fileData: result.data
       })
     })
+    // User retrieves the tutorial for the first time
```
Not sure what this comment is referring to?

Arcadius19 replied:
I accidentally left this comment when I rewrote the code.

NickWu007 commented on db/db-populate.js:
```diff
@@ -0,0 +1,9 @@
+const tutorialDB = require('../db/tutorial-queries')
+
+module.exports.populateDB = function () {
+  for (let i = 0; i < 2; i++) {
```
I think this the code you run when the table is first set up? Maybe make a note on the role of this.

NickWu007 commented on db/file-queries.js:
```diff
@@ -5,7 +5,13 @@ var db = require('./db-connect')
  */
 function getFileForUser (username, tutorialId) {
-  return db.one('select "data" from "LinksFile" where "username" = $1 and "tutorial_id"=$2', [username, tutorialId])
+  return db.task('getFileForUSer', t => {
```
Typo: `getFileForUSer` should be `getFileForUser`

NickWu007 commented on db/file-queries.js:
```diff
@@ -5,7 +5,13 @@ var db = require('./db-connect')
+    return db.oneOrNone('SELECT "data" FROM "LinksFile" WHERE "username" = $1 AND "tutorial_id"=$2', [username, tutorialId])
+      .then(file => {
+        return file || t.one('INSERT INTO "LinksFile" ("data", "tutorial_id", "username") VALUES' +
+          '((SELECT "source" FROM "LinksTutorial" WHERE "tutorial_id" = $1), $1, $2) RETURNING "data"', [tutorialId, username])
```
Since you are only returning the `data` field from DB, instead of the whole row, this function should be renamed into something like `getFileSourceForUser` or `getTutorialSourceForUser`.

NickWu007 commented on db/tutorial-queries.js:
```diff
@@ -0,0 +1,24 @@
+}
+
+function getHeaders () {
+  return db.many('SELECT tutorial_id, title FROM "LinksTutorial" ORDER BY tutorial_id ASC')
```
`tutorial_id` and `title` should be in quotes, so that they match the rest of the codebase.

NickWu007 commented on tutorials/descriptions.js:
```diff
@@ -0,0 +1,81 @@
+module.exports.startingLinksTitles = [
```
Again this seems to be an initial setup state of the tutorials. I am confused in why you have 6 titles and only 2 descriptions.

Also I think these "helper" files can be moved to a different directory to avoid confusion.

NickWu007 commented on api/compile.js:
```diff
@@ -3,12 +3,14 @@ const pf = require('portfinder')
 const fileDB = require('../db/file-queries')
 const { spawn } = require('child_process')
 
+module.exports.sessionMap = new Map();
```
No semicolon. Please EsLint this.

NickWu007 commented on api/reset.js:
```diff
@@ -1,6 +1,6 @@
 var fileDB = require('../db/file-queries')
 
-const startSources = require('./sources').startingLinksSources
+const startSources = require('../tutorials/sources').startingLinksSources
```
Instead of using this, you should get it from DB.

Arcadius19 replied:
This file was automatically changed when I moved sources.js file to a different directory. Actually, we do not need a reset API because it was only used to automatically populate LinksFile table for a newly created user, so I think it can be deleted.

NickWu007 replied:
In that case you can remove it, as long as we are populating the tutorial files properly.

Arcadius19 committed:
Refactored the code to be consistent with  ESLint configuration and the rest of the code.
Removed reset.js since the LinksFile entries are created upon a request for the tutorial for the first time instead of the user creation.

NickWu007 commented:
LGTM. Please hold on till this afternoon's meeting before merging this. Thanks.

Arcadius19 commented:
No problem

NickWu007 commented:
Going to merge this for now, as agreed.
